flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1286,10 +1289,38 @@ private void transferAllDataFromStateHandles(
 			Map<StateHandleID, StreamStateHandle> stateHandleMap,
 			Path restoreInstancePath) throws IOException {
 
+			ExecutorService executorService = Executors.newFixedThreadPool(5);
+			List<FutureTask<Void>> tasks = new ArrayList<>(stateHandleMap.size());
+
 			for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
 				StateHandleID stateHandleID = entry.getKey();
 				StreamStateHandle remoteFileHandle = entry.getValue();
-				copyStateDataHandleData(new Path(restoreInstancePath, stateHandleID.toString()), remoteFileHandle);
+
+				FutureTask<Void> task = new FutureTask<>(() -> {
```

Comment:
I think it would be more maintainable if we extract these 3 methods:
- `transferAllStateDataToDirectory`
- `transferAllDataFromStateHandles`
- `copyStateDataHandleData`
into a separate class file, e.g. `RocksDbStateDataTransfer`.
`RocksDBKeyedStateBackend` already contains a lot.